crates/tracing-instrumentation/src/exporter.rs

```diff
@@ -137,3 +137,128 @@ impl<T: opentelemetry_sdk::export::trace::SpanExporter + 'static>
+    A: SpanExporter + 'static,
+    B: SpanExporter + 'static,
+{
+    fn export(
+        &mut self,
+        batch: Vec<SpanData>,
+    ) -> BoxFuture<'static, opentelemetry_sdk::export::trace::ExportResult> {
+        let runtime = Arc::clone(&self.runtime);
+        let services = self.services.clone();
+
+        // only build services traces if services exporter is set
+        let (services_batch, runtime_batch): (Vec<_>, Vec<_>) = batch
+            .into_iter()
+            .partition(|s| s.attributes.iter().any(|kv| kv.key == USER_SERVICE_KEY));
+
+        Box::pin(async move {
+            {
+                let mut a = match runtime.lock() {
+                    Ok(a) => a,
+                    Err(_) => return Err(TraceError::Other("exporter mutex is poisoned".into())),
+                };
+                a.export(runtime_batch)
+            }
+            .await?;
+
+            if services_batch.is_empty() {
+                return Ok(());
+            }
+
+            {
+                if let Some(services) = services {
+                    let mut services = match services.lock() {
+                        Ok(b) => b,
+                        Err(_) => {
+                            return Err(TraceError::Other("exporter mutex is poisoned".into()))
+                        }
+                    };
+                    services.export(services_batch)
+                } else {
+                    Box::pin(async { Ok(()) })
+                }
+            }
+            .await
+        })
+    }
+
+    fn force_flush(
+        &mut self,
+    ) -> BoxFuture<'static, opentelemetry_sdk::export::trace::ExportResult> {
+        let runtime = Arc::clone(&self.runtime);
+        let services = self.services.clone();
+        Box::pin(async move {
+            {
+                let mut runtime = match runtime.lock() {
+                    Ok(runtime) => runtime,
+                    Err(_) => return Err(TraceError::Other("exporter mutex is poisoned".into())),
+                };
+                runtime.force_flush()
+            }
+            .await?;
+
+            {
+                if let Some(services) = services {
+                    let mut services = match services.lock() {
+                        Ok(services) => services,
+                        Err(_) => {
+                            return Err(TraceError::Other("exporter mutex is poisoned".into()))
+                        }
+                    };
+                    services.force_flush()
+                } else {
+                    Box::pin(async { Ok(()) })
+                }
+            }
+            .await
+        })
+    }
+
+    fn set_resource(&mut self, resource: &Resource) {
+        if let Ok(mut a) = self.runtime.lock() {
+            a.set_resource(resource);
+        }
+
+        if let Some(ref services) = self.services {
+            if let Ok(mut b) = services.lock() {
+                b.set_resource(resource)
+            }
+        }
+    }
```

Comment:
I wonder how contended that lock is, do you know when is this called? I want to make sure we understand the performance implications of locking, cloning, and allocations in the exporter. I'm not sure how frequently export is called and whether it's in the synchronous path of tracing or if it's non-blocking.

Reply:
it is non blocking of tracing - traces are batched by the spanprocessor and those batches end up being exported. when last i checked, i concluded that export is never called concurrently by otel, but they dont exclude the possibility in the type system sadly. so i would expect this to be uncontended

Reply:
I think this can easily be rewritten in a way to avoid locking by using channels and a producer/consumer  pattern.